File: frontend/src/types/game.ts
export interface GamePreset {
  id: number;
  title: string;
  description: string;
  worldview: string;
  character_options: string[];
  max_attribute_points: number;
  is_custom?: boolean;
  attributes: PresetAttributeOption[];
}

export interface PresetAttributeOption {
  key: string;
  label: string;
  min_value: number;
  max_value: number;
  default_value: number;
}

export interface StartGamePayload {
  preset_id: number;
  selected_character_setting?: string | null;
  custom_worldview?: string | null;
  custom_character_setting?: string | null;
  allocated_attributes: Record<string, number>;
  custom_prompt?: string | null;
}

export interface NextGamePayload {
  session_id: number;
  user_choice: string;
}

export interface GameStepResponse {
  session_id: number;
  event: string;
  event_segments: string[];
  current_stats: Record<string, number>;
  is_ended: boolean;
  world_entry_limit: number;
  world_entries_used_today: number;
  model_call_limit: number;
  model_calls_used_today: number;
  next_choices?: string[];
  end_reason?: string | null;
  end_summary?: string | null;
}

export interface ForceExitPayload {
  session_id: number;
}

export interface ForceExitResponse {
  success: boolean;
  end_reason?: string | null;
  end_summary?: string | null;
}

export interface GameHistorySession {
  id: number;
  user_id: number;
  preset_id: number;
  current_stats: Record<string, number>;
  event_history: Array<{
    role: 'system' | 'assistant' | 'user';
    content: string;
    effects?: Record<string, number>;
  }>;
  is_ended: boolean;
}

export interface StoryEvent {
  id: number;
  role: 'system' | 'assistant' | 'user';
  content: string;
}

File: app/schemas/game.py
from pydantic import BaseModel, ConfigDict, Field


class GameStartRequest(BaseModel):
    preset_id: int = Field(gt=0)
    selected_character_setting: str | None = Field(default=None, max_length=80)
    custom_worldview: str | None = Field(default=None, max_length=2000)
    custom_character_setting: str | None = Field(default=None, max_length=500)
    allocated_attributes: dict[str, int] = Field(default_factory=dict)
    custom_prompt: str | None = Field(default=None, max_length=200)


class GameNextRequest(BaseModel):
    session_id: int = Field(gt=0)
    user_choice: str = Field(min_length=1, max_length=500)


class GameSessionRead(BaseModel):
    id: int
    user_id: int
    preset_id: int
    current_stats: dict[str, object]
    event_history: list[dict[str, object]]
    is_ended: bool

    model_config = ConfigDict(extra="forbid")


class PresetRead(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: int
    title: str
    description: str
    worldview: str
    character_options: list[str]
    max_attribute_points: int
    is_custom: bool = False
    attributes: list[dict[str, object]]


class GameStepResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    session_id: int
    event: str
    event_segments: list[str]
    current_stats: dict[str, object]
    is_ended: bool
    world_entry_limit: int
    world_entries_used_today: int
    model_call_limit: int
    model_calls_used_today: int
    next_choices: list[str]
    end_reason: str | None = None
    end_summary: str | None = None


class GameForceExitRequest(BaseModel):
    session_id: int = Field(gt=0)


class GameForceExitResponse(BaseModel):
    success: bool
    end_reason: str | None = None
    end_summary: str | None = None
